### ISSUE TITLE
question about gradient calculation with respect to weight

### ISSUE THREAD

**Issue by jay2002** [wy1iu/sphereface#2]
In Margin Inner Product, the gradient with respect to weight is very simple:

```
  // Gradient with respect to weight
  if (this->param_propagate_down_[0]) {
    caffe_cpu_gemm<Dtype>(CblasTrans, CblasNoTrans, N_, K_, M_, (Dtype)1.,
        top_diff, bottom_data, (Dtype)1., this->blobs_[0]->mutable_cpu_diff());
  }
```

But in large margin softmax, the gradient calcuation is much more complex...

Can you please tell me how to simplify the gradient calcuation?
I failed to derive it ...

**Comment by pangyupo** [comment]
same question here.
also in the forward function, the weight is directly normalized by overwriting:
`  /************************* normalize weight *************************/
  Dtype* norm_weight = this->blobs_[0]->mutable_cpu_data();
  Dtype temp_norm = (Dtype)0.;
  for (int i = 0; i < N_; i++) {
  	temp_norm = caffe_cpu_dot(K_, norm_weight + i * K_, norm_weight + i * K_);
  	temp_norm = (Dtype)1./sqrt(temp_norm);
  	caffe_scal(K_, temp_norm, norm_weight + i * K_);
  }`
shouldn't we keep the weight, but use the normlized_weight in the following computation, so we can actually do back propagation wrt to weight.

**Comment by pangyupo** [comment]
same question here. why update of weight it the same for m=1, 2, 3, 4

**Comment by pangyupo** [comment]
also, since you normalize the weight by overwriting in forward() instead of keeping the original weight like [here:](https://github.com/openai/weightnorm/blob/master/tensorflow/nn.py)
```
  Dtype* norm_weight = this->blobs_[0]->mutable_cpu_data();
  Dtype temp_norm = (Dtype)0.;
  for (int i = 0; i < N_; i++) {
  	temp_norm = caffe_cpu_dot(K_, norm_weight + i * K_, norm_weight + i * K_);
  	temp_norm = (Dtype)1./sqrt(temp_norm);
  	caffe_scal(K_, temp_norm, norm_weight + i * K_);
  }
```
so everytime the after update of parameter: weight = weight + grad_w,  the weight is simply `clipped` into the normalized version?

**Comment by tornadomeet** [comment]
+1, backward seems not correspond with forward.

**Comment by wy1iu** [comment]
It is actually a normalized version of the gradient, which can help converge more stably. The direction is the same as before. What we do here is simply to rescale the gradient. Similar idea and intuition also appear in https://arxiv.org/pdf/1707.04822.pdf. 

However, if you use the original gradient to do the backprop, you could still make it work and obtain similar results, but not as stable as this normalized one.
